src/libraries/System.IO.Pipelines/ref/System.IO.Pipelines.cs

```diff
@@ -95,10 +95,12 @@ public static partial class StreamPipeExtensions
     public partial class StreamPipeReaderOptions
     {
         public StreamPipeReaderOptions(System.Buffers.MemoryPool<byte>? pool = null, int bufferSize = -1, int minimumReadSize = -1, bool leaveOpen = false) { }
+        public StreamPipeReaderOptions(System.Buffers.MemoryPool<byte>? pool, int bufferSize, int minimumReadSize, bool leaveOpen, bool waitForData) { }
         public int BufferSize { get { throw null; } }
         public bool LeaveOpen { get { throw null; } }
         public int MinimumReadSize { get { throw null; } }
         public System.Buffers.MemoryPool<byte> Pool { get { throw null; } }
+        public bool WaitForData { get { throw null; } }
```

Comment:
Do we expect developers to want both the false and true behaviors?  I'm wondering if it's just a bug that the behavior is equivalent to false today and we could change it without adding a new property/ctor arg for it, or if there are legitimately cases where WaitForData = false is desired (either for functionality or perf).

Reply:
I think when the Stream isn't a "connected stream" (PipeStream, NetworkStream) is the biggest use case for false. If you're quickly consuming all of the entire buffer then the zero byte read is additional overhead.

Reply:
Is that a real use case? In what situation does someone create a PipeReader from a Stream that's not connected, and does the extra read overhead matter in such cases?  If it's actually something folks do and it's important, ok.  Just seems odd to me, but I'm happy to be educated otherwise.

Reply:
Yes, it's a real use case. `PipeReader` is used for buffering and parsing things that don't all come in a single read. While the ASP.NET Core use cases are all connected, it is used for other things and I wouldn't want to have this on by default.

Reply:
Ok.  FWIW, I think it'd be worth measuring the real impact on such a use case before going in this direction.  If nothing else it'd be good to understand what opting-in costs.

Reply:
Sure, the worse case scenario I think would be something like `FileStream` without an internal buffer on windows. The overhead of each async read is pretty high. I'll do a quick benchmark.

Reply:
> I think would be something like FileStream without an internal buffer on windows. The overhead of each async read is pretty high

Fair enough.  (Though a) @adamsitnik is currently working to make that particular example much faster and b) FileStream reading from a disk file rather than a pipe could special-case 0 to return immediately (if it doesn't already).

Reply:
Right but this still supports netstandard2.0.